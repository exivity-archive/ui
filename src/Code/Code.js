import PropTypes from 'prop-types'
import styled from 'styled-components'
import defaultTheme from '../theme/defaultTheme'

const Code = styled.code`
  font-family: ${props => props.theme.fonts.code.family};
  font-weight: ${props => props.theme.fonts.code.weight};
  
  background-color: ${props => props.theme.colours.mark};
  padding: 3px;
  border-radius: ${props => props.theme.border.radius};
`

Code.propTypes = {
  children: PropTypes.node.isRequired,
  theme: PropTypes.object
}

Code.defaultProps = {
  theme: defaultTheme
}

export default Code
