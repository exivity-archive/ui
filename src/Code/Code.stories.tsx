import React from 'react'
import { storiesOf } from '@storybook/react'

import { Paragraph } from '../Paragraph'

import { Code } from '.'

const code = `import { CodeBlock } from '@exivity/ui'

<CodeBlock language='jsx'>{this}</CodeBlock>`

storiesOf('atoms|Code', module)
  .add('default', () => <Paragraph>This is some <Code>inline code</Code> right here.</Paragraph>)
  .add('block', () => <Code block language='jsx'>{code}</Code>)
  .add('non-existing', () => (
    <Code block language='exivity-js'>
    Unsupported language displayed as text
    </Code>
  ))
