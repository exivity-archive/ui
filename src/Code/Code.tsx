import React, { ReactNode, useEffect, useState } from 'react'
// @ts-ignore
import SyntaxHighlighter from 'react-syntax-highlighter/dist/esm/prism-light'
// @ts-ignore
import solarizedlight from 'react-syntax-highlighter/dist/esm/styles/prism/solarizedlight'
import styled, { css } from 'styled-components'
import { BlockProps, fromTheme } from '../utils/styled'
import { preciseEm } from '../utils/styled/isolated'

interface CodeProps extends SyntaxHighlighterProps {
  language?: string
  block?: boolean
  showLineNumbers?: boolean
  children?: ReactNode
}

const useDynamicLanguage = (language: string) => {
  // Set initial state to a boolean indicating whether the language is text
  // (in which case we can skip the dynamic import effect)
  const [current, setCurrent] = useState('text')

  useEffect(() => {
    if (language !== 'text') {
      import(`react-syntax-highlighter/dist/esm/languages/prism/${language}`)
        .then(definition => {
          SyntaxHighlighter.registerLanguage(language, definition.default)
          setCurrent(language)
        })
        .catch(() => {
          console.error(`Language '${language}' could not be found, falling back to plaintext`)
          setCurrent('text')
        })
    }
  }, [language])

  return current
}

const StyledSyntaxHighlighter = styled(SyntaxHighlighter)`
  // Override globalBlockSpacing because we need !important
  margin: ${(props: BlockProps) => props.noMargin ? 0 : preciseEm(1)}rem 0 !important;

  &:first-child {
    margin-top: 0 !important;
  }

  &:last-child {
    margin-bottom: 0 !important;
  }

  & {
    background-color: ${fromTheme(theme => theme.colours.lightGray)} !important;
    &, code, pre {
      font-family: ${fromTheme(theme => theme.global.fontFamilyMonospace)} !important;
    }

    padding: ${props => props.PreTag === 'span'
      ? css`${fromTheme(theme => theme.global.spacing / 4)}`
      : css`${fromTheme(theme => theme.global.spacing)}`}em !important;
  }
`

export const Code = ({ children, block, showLineNumbers, language = 'text' }: CodeProps & BlockProps) => {
  const highlighterLanguage = useDynamicLanguage(language)

  return <StyledSyntaxHighlighter
    PreTag={block ? 'pre' : 'span'}
    style={solarizedlight}
    language={highlighterLanguage}
    showLineNumbers={showLineNumbers}
  >
    {children}
  </StyledSyntaxHighlighter>
}
