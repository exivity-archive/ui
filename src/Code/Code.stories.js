import React from 'react'
import { storiesOf } from '@storybook/react'

import Code from './Code'

storiesOf('atoms|Code', module)
  .add('default', () => <Code>Component</Code>)
