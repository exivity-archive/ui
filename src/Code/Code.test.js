/* eslint-disable no-undef */
import React from 'react'
import renderer from 'react-test-renderer'

import Code from './Code'

test('renders basic Code', () => {
  const component = renderer.create(<Code />)
  expect(component.toJSON()).toMatchSnapshot()
})
