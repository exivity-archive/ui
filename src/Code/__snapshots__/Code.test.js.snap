// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders basic Code 1`] = `
.c0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

<code
  className="c0"
/>
`;
